backend/dal/Services/Concrete/UserService.cs

```diff
@@ -107,6 +153,61 @@ public AccessRequest AddAccessRequest(AccessRequest request)
             return request;
         }
 
+        /// <summary>
+        /// Update the access request for the current user.
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        public AccessRequest UpdateAccessRequest(AccessRequest request)
+        {
+            if (request == null || request.Agencies == null || request.Roles == null) throw new ArgumentNullException(nameof(request));
+            var userId = this.User.GetUserId();
+            var position = request.User.Position;
+            request.User = this.Context.Users.Find(userId) ??
+                throw new KeyNotFoundException("Your account has not been activated.");
+
+            if (request.UserId != userId) throw new NotAuthorizedException(); // Not allowed to update someone elses request.
+
+            // fetch the existing request from the datasource.
+            var entity = this.Context.AccessRequests
+                .Include(a => a.Agencies)
+                .ThenInclude(a => a.Agency)
+                .Include(a => a.Roles)
+                .ThenInclude(r => r.Role)
+                .Include(a => a.User)
+                .FirstOrDefault(a => a.Id == request.Id) ?? throw new KeyNotFoundException();
+
+            if (entity.User.Position != position)
+            {
+                entity.User.Position = position;
+                this.Context.Entry(entity.User).State = EntityState.Modified;
+            }
+
+            // Remove agencies and roles if required.
+            var removeAgencies = entity.Agencies.Except(request.Agencies, new AccessRequestAgencyAgencyIdComparer());
+            if (removeAgencies.Any()) entity.Agencies.RemoveAll(a => removeAgencies.Any(r => r.AgencyId == a.AgencyId));
+
+            var removeRoles = entity.Roles.Except(request.Roles, new AccessRequestRoleRoleIdComparer());
+            if (removeRoles.Any()) entity.Roles.RemoveAll(a => removeRoles.Any(r => r.RoleId == a.RoleId));
+
+            // Add agencies and roles if required.
+            var addAgencies = request.Agencies.Except(entity.Agencies, new AccessRequestAgencyAgencyIdComparer());
+            addAgencies.ForEach(a => entity.Agencies.Add(a));
+
+            var addRoles = request.Roles.Except(entity.Roles, new AccessRequestRoleRoleIdComparer());
+            addRoles.ForEach(r => entity.Roles.Add(r));
+
+            // Copy values into entity.
+            this.Context.Entry(entity).CurrentValues.SetValues(request);
+            this.Context.SetOriginalRowVersion(entity);
+
+            this.Context.AccessRequests.Update(entity);
+            this.Context.CommitTransaction();
+            this.Context.Entry(request).CurrentValues.SetValues(entity);
```

Comment:
I want to mutate the `request` as it is our common pattern everywhere else.  If I don't do this the `request` will not have the latest `rowVersion` and as such the subsequent request to update will fail.

Reply:
Does this do anything though? you mutate the request value but don't return it. Aren't we expecting that a consumer will use the returned entity rather then a mutated parameter?

Reply:
This was bugging me with our current implementation.  I generally feel we shouldn't mutate objects if we can avoid it, however in the majority (possibly all) our current uses we expect the entity to be mutated.  

This particular case was complicated by the fact that we need to pull the latest data from the DB to determine what has changed.  EF will get confused if we try and track both entities, which means I choose to manually mutate the `request` so that it wouldn't confuse the common use throughout the backend.

To stop mutating entities passed in would be a lot of effort and I'm not sure we gain much by doing that.

Reply:
To be clear, if I don't mutate then the `rowVersion` will not be updated.  In most of our endpoints we don't necessarily use the returned entity.  Which will result in the wrong data being returned.